include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -5275,6 +5278,165 @@ struct OrtApi {
+
+  /** \brief Returns information (OrtNode and input index) for all uses of the given OrtValueInfo as an input to
+   * an OrtNode instance.
+   *
+   * Caller provides 2 pre-allocated arrays that will be filled with the OrtNode and input index values.
+   * Use GetValueInfoNumUses() to get the number of uses of the OrtValueInfo.
+   *
+   * An OrtNode instance may appear multiple times if it uses the given OrtValueInfo more than once (e.g., Mul(x, x)).
+   *
+   * \param[in] OrtValueInfo The OrtValueInfo instance.
+   * \param[out] nodes Pre-allocated array of size `max_num_uses` that will be filled with OrtNode instances.
+   * \param[out] input_indices Pre-allocated array of `max_num_uses` elements that will be filled
+   *                           with input indices.
+   * \param[in] max_num_uses The maximum size of the `consumer_nodes` and `consumer_input_indices` arrays.
+                             Typical usage sets this to the value of GetValueInfoNumUses().
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   * \since Version 1.23.
+   */
+  ORT_API2_STATUS(GetValueInfoUses, _In_ const OrtValueInfo* value_info,
+                  _Out_writes_all_(max_num_uses) const OrtNode** nodes,
+                  _Out_writes_all_(max_num_uses) size_t* input_indices,
+                  _In_ size_t max_num_uses);
+
+  /** \brief Returns the name of a OrtGraph instance.
+   * \param[in] graph The OrtGraph instance.
+   * \return The OrtGraph instance's name.
+   *
+   * \since Version 1.23.
+   */
+  const char*(ORT_API_CALL* Graph_GetName)(_In_ const OrtGraph* graph);
+
+  /** \brief Returns the number of nodes in the OrtGraph instance.
+   * \param[in] graph The OrtGraph instance.
+   * \return The number of nodes.
+   *
+   * \since Version 1.23.
+   */
+  size_t(ORT_API_CALL* Graph_GetNumNodes)(_In_ const OrtGraph* graph);
+
+  /** \brief Returns an array of the OrtNode instances contained in the OrtGraph.
+   *
+   * Caller provides a pre-allocated array that will be filled with the nodes. Use OrtGraph_NumNodes() to get the
+   * number of nodes in the graph. The nodes are sorted according to the `order` argument.
+   *
+   * \param[in] graph The OrtGraph instance.
+   * \param[in] order The ordering of the nodes:
+   *                  0 means the nodes are sorted in topological order.
+   *                  1 means the nodes are sorted in topological order with priority.
+   *                  2 means the nodes are sorted in memory efficient topological order.
+   * \param[out] nodes Pre-allocated array of `max_num_nodes` elements that will be filled with OrtNode pointers.
+   * \param[in] max_num_nodes The maximum size of the nodes array.
+                              Typical usage sets this to the value of OrtGraph_NumNodes.
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   * \since Version 1.23.
+   */
+  ORT_API2_STATUS(Graph_GetNodes, const OrtGraph* graph, int order,
```

Comment:
Questions:
1. Should we even expose the `order` parameter? All values of order return some sort of topological ordering. I think some EPs, like TensorRT EP, currently request "topological order with priority", but is that truly needed? @chilo-ms 
2. Instead of returning all nodes in an array, should we consider returning an iterator?
   - Intel noted that [OpenVINO's compiler frontend uses a graph iterator](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/common/include/openvino/frontend/graph_iterator.hpp#L15).
   - ONNX IR also represents a [GraphProtocol as an iterable sequence of nodes](https://github.com/onnx/ir-py/blob/main/src/onnx_ir/_protocols.py#L266).

Reply:
> I think some EPs, like TensorRT EP, currently request "topological order with priority", but is that truly needed?

Yes, TRT EP requires "topological order with priority" (it's not the default in ORT) to [add ](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/graph/graph_proto_serializer.cc#L57)node proto to graph proto in that order when doing GraphViewerToProto. And passing the final ModelProto/GraphProto to TRT parser.
We did see that, in the case of ResNet50, if using ORT's default topo sort, the GraphProto has different node ordering and will end up making TRT parse get the wrong node number when deserializing.
Related PR https://github.com/microsoft/onnxruntime/pull/20411

Reply:
I do think it feels EP-specific. Additionally, an EP should be able to use the ORT graph IR itself to build their own topological ordering of nodes. ONNX IR, for example, also [requires the API user to do their own sort](https://github.com/onnx/ir-py/blob/de50b14d31aeb73bcf11cda867a3869b6849e6a3/src/onnx_ir/_protocols.py#L274-L276):

>The `nodes` are not guaranteed to be topologically sorted. But the iteration order should be deterministic across different runs. It is the responsibility of the user to maintain a topological order of the nodes.

That said, given the tight timelines, I don't know if we want to force EPs to write their own sort as part of the transition. We can chat offline.

Reply:
Agree that it feels EP-specific.
My concern is also the timelines and for ONNX IR, even though it mentioned that "It is the responsibility of the user to maintain a topological order of the nodes", it still provides sort() for the GraphProtocol.

Reply:
The priority-based topological sort is actually not very much code. I was able to write this same sorting algorithm using only the new public C API graph types. I added a unit test that loads a sample model, runs the sort, and compares with ORT's built-in sort.

Check it out:

https://github.com/microsoft/onnxruntime/blob/a9d8956b4dd791dc345cd932b06697ba16729b85/onnxruntime/test/ep_graph/test_ep_graph.cc#L375-L436

https://github.com/microsoft/onnxruntime/blob/a9d8956b4dd791dc345cd932b06697ba16729b85/onnxruntime/test/ep_graph/test_ep_graph.cc#L472-L492

Based on the above, I'm strongly considering removing the `order` parameter and just returning nodes in our default topo order. @chilo-ms

Reply:
Removed the `order` parameter for now.